File: python/prism/protocol.py
# protocol.py — PRISM binary frame encode/decode
import struct
from .constants import (
    SRV_OK, SRV_BUSY, SRV_BAD_PARAM,
    RESULT_SUCCESS, RESULT_BUSY, RESULT_BAD_PARAMETER,
    RESULT_COMMS_FAILURE, RESULT_UNDEFINED,
    CMD_SCAN_DATA, SCAN_BYTES_PER_S,
)


def build_req(cmd: int, payload: bytes = b"") -> bytes:
    """[cmd:1][len:1][payload:N]"""
    return bytes([cmd, len(payload)]) + payload


def parse_res_hdr(hdr: bytes, expected_cmd: int) -> tuple[int, int]:
    """Parse 3-byte response header → (result_code, payload_len).
    Raises ValueError on unexpected cmd."""
    cmd, status, payload_len = hdr[0], hdr[1], hdr[2]
    if cmd != expected_cmd:
        raise ValueError(
            f"Unexpected response cmd 0x{cmd:02X} (expected 0x{expected_cmd:02X})"
        )
    return srv_status_to_result(status), payload_len


def srv_status_to_result(status: int) -> int:
    if status == SRV_OK:        return RESULT_SUCCESS
    if status == SRV_BUSY:      return RESULT_BUSY
    if status == SRV_BAD_PARAM: return RESULT_BAD_PARAMETER
    return RESULT_UNDEFINED


def int24_to_int32(data: bytes, offset: int) -> int:
    """3-byte little-endian int24 → signed int32."""
    v = data[offset] | (data[offset + 1] << 8) | (data[offset + 2] << 16)
    if v & 0x800000:
        v -= 0x1000000
    return v


def pack_double(value: float) -> bytes:
    return struct.pack("<d", value)


def unpack_double(data: bytes, offset: int = 0) -> float:
    return struct.unpack_from("<d", data, offset)[0]


def unpack_i32(data: bytes, offset: int = 0) -> int:
    return struct.unpack_from("<i", data, offset)[0]


def is_scan_frame(frame: bytes) -> bool:
    return len(frame) >= 1 and frame[0] == CMD_SCAN_DATA


def scan_frame_status(frame: bytes) -> int:
    return frame[2]


def scan_frame_payload(frame: bytes) -> bytes:
    return frame[3:]


def decode_scan_payload(payload: bytes, n_samples: int, ch_count: int,
                        cal_slope: list, cal_offset: list,
                        sensitivity: list, options: int) -> list[float]:
    """Convert raw 24-bit ADC payload to physical doubles.
    Returns interleaved list: [ch0_s0, ch1_s0, ..., ch0_s1, ...]
    """
    from .constants import ScanOptions
    out: list[float] = []
    pos = 0
    for _ in range(n_samples):
        for ch in range(ch_count):
            raw = int24_to_int32(payload, pos)
            pos += SCAN_BYTES_PER_S

            code = float(raw)
            if not (options & ScanOptions.NOCALIBRATEDATA):
                code = (code - cal_offset[ch]) * cal_slope[ch]

            if options & ScanOptions.NOSCALEDATA:
                out.append(code)
            else:
                voltage = code / 8_388_607.0 * 5.0
                out.append(voltage / (sensitivity[ch] / 1000.0))
    return out

File: lib/protocol.c
/*
 * protocol.c — PRISM binary frame encode/decode
 *
 * Wire formats (all little-endian):
 *   Request : [cmd:1][payload_len:1][payload:payload_len]
 *   Response: [cmd:1][status:1][payload_len:1][payload:payload_len]
 *   ScanPush: [CMD_SCAN_DATA:1][n_samples:1][status:1][768-byte ADC data]
 */
#include "protocol.h"
#include "../include/prismc.h"

#include <string.h>

/* ── Pack helpers ───────────────────────────────────────────────────────── */

void proto_pack_u8(uint8_t *p, uint8_t v)   { p[0] = v; }
void proto_pack_u32(uint8_t *p, uint32_t v) { memcpy(p, &v, 4); }
void proto_pack_double(uint8_t *p, double v) { memcpy(p, &v, 8); }

uint8_t proto_unpack_u8(const uint8_t *p)    { return p[0]; }
int32_t proto_unpack_i32(const uint8_t *p)   { int32_t v; memcpy(&v, p, 4); return v; }
double  proto_unpack_double(const uint8_t *p){ double  v; memcpy(&v, p, 8); return v; }

/* Sign-extend 3-byte little-endian int24 to int32 */
int32_t proto_int24_to_int32(const uint8_t *p)
{
    int32_t v = (int32_t)((uint32_t)p[0]
                         | ((uint32_t)p[1] << 8)
                         | ((uint32_t)p[2] << 16));
    if (v & 0x800000)
        v |= (int32_t)0xFF000000;
    return v;
}

/* ── Request builder ────────────────────────────────────────────────────── */

int proto_build_req(uint8_t cmd, const uint8_t *payload, uint8_t payload_len,
                    uint8_t *out_buf)
{
    out_buf[0] = cmd;
    out_buf[1] = payload_len;
    if (payload_len > 0 && payload)
        memcpy(&out_buf[2], payload, payload_len);
    return 2 + payload_len;
}

/* ── Response header parser ─────────────────────────────────────────────── */

int proto_parse_res_hdr(const uint8_t hdr[3], uint8_t expected_cmd,
                        int *result_out, uint8_t *payload_len_out)
{
    if (hdr[0] != expected_cmd) {
        *result_out = RESULT_COMMS_FAILURE;
        return -1;
    }
    *result_out      = proto_srv_status_to_result(hdr[1]);
    *payload_len_out = hdr[2];
    return (int)hdr[2];
}

/* ── Scan frame helpers ─────────────────────────────────────────────────── */

int proto_is_scan_frame(const uint8_t *frame)
{
    return frame[0] == (uint8_t)CMD_SCAN_DATA;
}

uint8_t proto_scan_frame_status(const uint8_t *frame)
{
    return frame[2];
}

const uint8_t *proto_scan_frame_payload(const uint8_t *frame)
{
    return &frame[3];
}

/* ── Status mapping ─────────────────────────────────────────────────────── */

int proto_srv_status_to_result(uint8_t s)
{
    switch (s) {
        case SRV_OK:        return RESULT_SUCCESS;
        case SRV_BUSY:      return RESULT_BUSY;
        case SRV_BAD_PARAM: return RESULT_BAD_PARAMETER;
        default:            return RESULT_UNDEFINED;
    }
}
